### PR TITLE
convert circle to use be an implmentation of buffer2

### PR DESCRIPTION
**Pull request by mcwhittemore**
@lucaswoj All the tests are passing and my maps look right. Can you confirm this looks right to you before I start working on the rest?

### CODE REVIEW

File: js/data/buffer/circle_vertex_buffer.js [hunk 1]
```diff
@@ -1,41 +1,31 @@
...
- * This contains the data that displays circle markers on the map,
- * including their centerpoint
- */
-function CircleVertexBuffer(buffer) {
-    Buffer.call(this, buffer);
+var Buffer2 = require('../buffer2');
+
+function CircleVertexBuffer(options) {
+    Buffer2.call(this, options || {
+        type: Buffer2.BufferType.ELEMENT,
```
Comment (lucaswoj) [inline js/data/buffer/circle_vertex_buffer.js]: This should be `Buffer2.BufferType.VERTEX`. 

There are two kinds of buffers: element buffers and vertex buffers. 

 - Vertex buffers encode a set of user-defined attributes per vertex.
 - Element buffers encode references to sets of three vertices that constitute a triangle. 

(Apologies if this is review for you!)

**Comment by lucaswoj** [comment]
Other than my one comment above, this looks great!

File: js/data/buffer/circle_vertex_buffer.js (hunk 1 above)
Comment (mcwhittemore) [inline js/data/buffer/circle_vertex_buffer.js]: So, element is only for triangle? The rest will be vertex?

**Comment by mcwhittemore** [comment]
Tests seem to be inconsistent... 

https://circleci.com/gh/mapbox/mapbox-gl-js/954
https://circleci.com/gh/mapbox/mapbox-gl-js/955

File: js/data/buffer/circle_vertex_buffer.js (hunk 1 above)
Comment (lucaswoj) [inline js/data/buffer/circle_vertex_buffer.js]: The class name indicates what type of buffer it is. `CircleVertexBuffer` is a vertex buffer. Every call to `draw` in GL requires both an element buffer and a vertex buffer. All shapes are fundamentally drawn in terms of triangles, even lines and circles.

**Comment by lucaswoj** [comment]
Huh. No idea what would cause a segfault on circleci. Have you been able to reproduce that locally?

File: js/data/buffer/circle_vertex_buffer.js [hunk 2]
```diff
@@ -1,41 +1,31 @@
...
-            pos2 = pos / 2;
-
-        this.resize();
-
-        // pack the extrusion of -1 or 1 into one short
-        this.shorts[pos2 + 0] = (x * 2) + ((extrudeX + 1) / 2);
-        this.shorts[pos2 + 1] = (y * 2) + ((extrudeY + 1) / 2);
-
-        this.pos += this.itemSize;
+        this.push([x, y, extrudeX, extrudeY]);
```
Comment (lucaswoj) [inline js/data/buffer/circle_vertex_buffer.js]: Oh -- looks like you'll want to change this to 

```
this.push([
  (x * 2) + ((extrudeX + 1) / 2),
  (y * 2) + ((extrudeY + 1) / 2)
]);
```

... that probably explains the segfault.

Comment (mcwhittemore) [inline js/data/buffer/circle_vertex_buffer.js]: components should be a 2 than too right?

Comment (lucaswoj) [inline js/data/buffer/circle_vertex_buffer.js]: Yes! Good catch :+1:

**Comment by mcwhittemore** [comment]
@lucaswoj should be ready to merge when you're ready.

**Comment by lucaswoj** [comment]
Squash 'er down into one commit and :ship: @mcwhittemore!

**Pull request merged**